File: src/communication/rs485_bridge.py
"""Dual-mode RS-485 bridge (TCP-to-serial-server or local serial port).

Supports two transport modes:

* **network** -- TCP connection to an RS-485-to-Ethernet serial server.
* **serial**  -- Direct local serial port (COM/ttyUSB/ttyACM, etc.).

Inbound data is split on CRLF delimiters and each line is emitted
via the :pyqtSignal:`data_received` signal.
"""

from PyQt6.QtCore import QObject, pyqtSignal
from PyQt6.QtNetwork import QTcpSocket, QAbstractSocket
import logging

logger = logging.getLogger(__name__)


class Rs485Bridge(QObject):
    """Dual-mode RS-485 communication bridge.

    Use :meth:`connect_network` for a TCP serial-server, or
    :meth:`connect_serial` for a local COM/tty port.
    """

    data_received = pyqtSignal(bytes)              # complete line (no CRLF)
    connection_state_changed = pyqtSignal(object)   # DeviceStatus or None

    def __init__(self, parent=None):
        super().__init__(parent)
        self._mode = "network"
        self._tcp_socket: QTcpSocket | None = None
        self._serial_port = None
        self._buffer = bytearray()

    # ------------------------------------------------------------------
    # Network mode (TCP serial server)
    # ------------------------------------------------------------------

    def connect_network(self, host: str, port: int):
        """Connect to an RS-485 serial server over TCP."""
        self.disconnect_all()
        self._mode = "network"
        self._tcp_socket = QTcpSocket(self)
        self._tcp_socket.readyRead.connect(self._on_tcp_ready_read)
        self._tcp_socket.connected.connect(self._emit_connected)
        self._tcp_socket.disconnected.connect(self._emit_disconnected)
        self._tcp_socket.errorOccurred.connect(self._on_tcp_error)
        self._tcp_socket.connectToHost(host, port)

    # ------------------------------------------------------------------
    # Serial mode (local COM/tty port)
    # ------------------------------------------------------------------

    def connect_serial(
        self,
        port_name: str,
        baud_rate: int = 9600,
        data_bits: int = 8,
        stop_bits: int = 1,
        parity: int = 0,
    ):
        """Open a local serial port for RS-485 communication.

        Args:
            port_name: Serial port name (e.g. ``COM3`` or ``/dev/ttyUSB0``).
            baud_rate: Baud rate (default 9600).
            data_bits: Data bits (default 8).
            stop_bits: Stop bits (default 1).
            parity: Parity (0=None, 1=Even, 2=Odd, 3=Mark, 4=Space).
        """
        from PyQt6.QtSerialPort import QSerialPort

        self.disconnect_all()
        self._mode = "serial"
        self._serial_port = QSerialPort(self)
        self._serial_port.setPortName(port_name)
        self._serial_port.setBaudRate(baud_rate)
        self._serial_port.setDataBits(data_bits)
        self._serial_port.setStopBits(stop_bits)
        self._serial_port.setParity(parity)
        self._serial_port.readyRead.connect(self._on_serial_ready_read)
        if not self._serial_port.open(QSerialPort.OpenModeFlag.ReadOnly):
            logger.error("Failed to open serial port %s", port_name)
            self._emit_disconnected()
        else:
            logger.info("RS-485 serial opened on %s @ %d", port_name, baud_rate)
            self._emit_connected()

    # ------------------------------------------------------------------
    # Common
    # ------------------------------------------------------------------

    def disconnect_all(self):
        """Close any open connection and clear buffers."""
        if self._tcp_socket:
            self._tcp_socket.disconnectFromHost()
            self._tcp_socket.deleteLater()
            self._tcp_socket = None
        if self._serial_port and self._serial_port.isOpen():
            self._serial_port.close()
            self._serial_port.deleteLater()
            self._serial_port = None
        self._buffer.clear()

    def send_data(self, data: bytes):
        """Send raw bytes to the RS-485 bus."""
        if self._mode == "network" and self._tcp_socket:
            self._tcp_socket.write(data)
        elif self._mode == "serial" and self._serial_port:
            self._serial_port.write(data)

    def is_connected(self) -> bool:
        """Return True if the current transport is open/connected."""
        if self._mode == "network" and self._tcp_socket:
            return (
                self._tcp_socket.state()
                == QAbstractSocket.SocketState.ConnectedState
            )
        if self._mode == "serial" and self._serial_port:
            return self._serial_port.isOpen()
        return False

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    def _emit_connected(self):
        from ..core.common.types import DeviceStatus
        self.connection_state_changed.emit(DeviceStatus.IDLE)

    def _emit_disconnected(self):
        from ..core.common.types import DeviceStatus
        self.connection_state_changed.emit(DeviceStatus.OFFLINE)

    def _on_tcp_error(self, error):
        logger.error(
            "RS-485 TCP error: %s", self._tcp_socket.errorString()
        )

    def _on_tcp_ready_read(self):
        data = self._tcp_socket.readAll().data()
        self._process_data(data)

    def _on_serial_ready_read(self):
        data = self._serial_port.readAll().data()
        self._process_data(data)

    def _process_data(self, data: bytes):
        """Split incoming data on CRLF boundaries and emit lines."""
        self._buffer.extend(data)
        while b'\r\n' in self._buffer:
            idx = self._buffer.index(b'\r\n')
            frame = bytes(self._buffer[:idx])
            self._buffer = self._buffer[idx + 2:]
            if frame:
                self.data_received.emit(frame)

File: src/communication/rs485/Rs485Bridge.cpp
#include "communication/rs485/Rs485Bridge.h"
#include <QDebug>

Rs485Bridge::Rs485Bridge(QObject* parent)
    : QObject(parent)
{
}

Rs485Bridge::~Rs485Bridge()
{
    disconnectNetwork();
    disconnectSerial();
}

// --- Network mode ---

bool Rs485Bridge::connectNetwork(const QString& host, int port)
{
    disconnectSerial();

    if (!m_tcpSocket) {
        m_tcpSocket = new QTcpSocket(this);
        connect(m_tcpSocket, &QTcpSocket::connected, this, &Rs485Bridge::onTcpConnected);
        connect(m_tcpSocket, &QTcpSocket::disconnected, this, &Rs485Bridge::onTcpDisconnected);
        connect(m_tcpSocket, &QTcpSocket::readyRead, this, &Rs485Bridge::onTcpReadyRead);
        connect(m_tcpSocket, &QTcpSocket::errorOccurred, this, &Rs485Bridge::onTcpError);
    }

    m_mode = NetworkMode;
    m_readBuffer.clear();

    qDebug() << "Rs485Bridge: Connecting to serial server at" << host << "port" << port;
    emit connectionStateChanged(DeviceStatus::Offline);
    m_tcpSocket->connectToHost(host, port);

    return true;
}

void Rs485Bridge::disconnectNetwork()
{
    if (m_tcpSocket) {
        if (m_tcpSocket->state() != QAbstractSocket::UnconnectedState) {
            m_tcpSocket->disconnectFromHost();
        }
        m_tcpSocket->deleteLater();
        m_tcpSocket = nullptr;
    }
    m_readBuffer.clear();
}

// --- Serial mode ---

bool Rs485Bridge::connectSerial(const QString& portName, int baudRate,
                                int dataBits, int stopBits, int parity)
{
    disconnectNetwork();

    if (!m_serialPort) {
        m_serialPort = new QSerialPort(this);
        connect(m_serialPort, &QSerialPort::readyRead, this, &Rs485Bridge::onSerialReadyRead);
        connect(m_serialPort, &QSerialPort::errorOccurred, this, &Rs485Bridge::onSerialError);
    }

    m_mode = SerialMode;
    m_readBuffer.clear();

    m_serialPort->setPortName(portName);
    m_serialPort->setBaudRate(baudRate);
    m_serialPort->setDataBits(static_cast<QSerialPort::DataBits>(dataBits));
    m_serialPort->setStopBits(static_cast<QSerialPort::StopBits>(stopBits));
    m_serialPort->setParity(static_cast<QSerialPort::Parity>(parity));

    qDebug() << "Rs485Bridge: Opening serial port" << portName
             << "baud=" << baudRate << "data=" << dataBits
             << "stop=" << stopBits << "parity=" << parity;

    emit connectionStateChanged(DeviceStatus::Offline);

    if (!m_serialPort->open(QIODevice::ReadOnly)) {
        qWarning() << "Rs485Bridge: Failed to open serial port" << portName
                   << "-" << m_serialPort->errorString();
        return false;
    }

    emit connectionStateChanged(DeviceStatus::Idle);
    return true;
}

void Rs485Bridge::disconnectSerial()
{
    if (m_serialPort) {
        if (m_serialPort->isOpen()) {
            m_serialPort->close();
        }
        m_serialPort->deleteLater();
        m_serialPort = nullptr;
    }
    m_readBuffer.clear();
}

// --- Send data ---

void Rs485Bridge::sendData(const QByteArray& data)
{
    if (m_mode == NetworkMode && m_tcpSocket &&
        m_tcpSocket->state() == QAbstractSocket::ConnectedState) {
        m_tcpSocket->write(data);
    } else if (m_mode == SerialMode && m_serialPort && m_serialPort->isOpen()) {
        m_serialPort->write(data);
    }
}

bool Rs485Bridge::isConnected() const
{
    if (m_mode == NetworkMode) {
        return m_tcpSocket && m_tcpSocket->state() == QAbstractSocket::ConnectedState;
    } else {
        return m_serialPort && m_serialPort->isOpen();
    }
}

// --- TCP slots ---

void Rs485Bridge::onTcpConnected()
{
    qDebug() << "Rs485Bridge: Connected to serial server (network mode)";
    m_readBuffer.clear();
    emit connectionStateChanged(DeviceStatus::Idle);
}

void Rs485Bridge::onTcpDisconnected()
{
    qDebug() << "Rs485Bridge: Disconnected from serial server (network mode)";
    emit connectionStateChanged(DeviceStatus::Offline);
}

void Rs485Bridge::onTcpReadyRead()
{
    m_readBuffer.append(m_tcpSocket->readAll());
    parseFrames();
}

void Rs485Bridge::onTcpError(QAbstractSocket::SocketError error)
{
    Q_UNUSED(error)
    qWarning() << "Rs485Bridge: TCP error:" << m_tcpSocket->errorString();
}

// --- Serial slots ---

void Rs485Bridge::onSerialReadyRead()
{
    m_readBuffer.append(m_serialPort->readAll());
    parseFrames();
}

void Rs485Bridge::onSerialError(QSerialPort::SerialPortError error)
{
    if (error != QSerialPort::NoError && error != QSerialPort::ResourceError) {
        qWarning() << "Rs485Bridge: Serial port error:" << m_serialPort->errorString();
    }

    if (error == QSerialPort::ResourceError) {
        qWarning() << "Rs485Bridge: Serial port resource error, port may have been removed";
        emit connectionStateChanged(DeviceStatus::Fault);
    }
}

// --- Frame parsing ---

void Rs485Bridge::parseFrames()
{
    // RS485 frame delimiter: 0x0D 0x0A (CRLF)
    while (true) {
        int delimiterPos = m_readBuffer.indexOf("\r\n");
        if (delimiterPos < 0) {
            break;
        }

        // Extract the frame data (everything before the delimiter)
        QByteArray frameData = m_readBuffer.left(delimiterPos);
        m_readBuffer.remove(0, delimiterPos + 2); // +2 for CRLF

        if (!frameData.isEmpty()) {
            emit dataReceived(frameData);
        }
    }

    // Safety: prevent unbounded buffer growth
    if (m_readBuffer.size() > 65536) {
        qWarning() << "Rs485Bridge: Read buffer overflow, clearing" << m_readBuffer.size() << "bytes";
        m_readBuffer.clear();
    }
}
